src/cuda_block_manager.h

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+#pragma once
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+#include <memory>
+#include <mutex>
+#include <unordered_map>
+#include <vector>
+
+#include "constants.h"
+#include "status.h"
+
+
+namespace triton { namespace core {
+
+class Allocation;
+
+///
+/// CudaBlockManager uses CUDA Virtual Memory Management APIs to allow the
+/// buffers to grow without having to create a new allocation. Currently,
+/// it is only used with implicit state management.
+///
+class CudaBlockManager {
+ public:
+  static Status Create(double min_supported_compute_capability);
```

Comment:
A few questions:
1) Generally for a singleton pattern - we would have a "GetInstance()" method that creates or returns the instance, and then all calls would be instance methods and not static methods. If we moved them to instance methods we could avoid the 'instance == null' check.
2) It seems odd that the Allocate() and Free() methods take different objects. Is there a reason why Free() wouldn't take an Allocation? 
3) Not sure I follow the reason the allocate takes a reference to a unique pointer - could it take a reference to an allocation instead?

Reply:
1. I agree but I followed the existing pattern that exists in `cuda_memory_manager`. I have filed a ticket to revisit this in the future.
2. I think we should be able to make that change.
3. It is just that the move semantics for `Allocation` are clearer in my head and I don't have to worry about the copies (since unique ptr doesn't allow copies). Happy to change it to a reference if you think it is less confusing.

Reply:
Used the same object for both Allocate and Free.